hastexo/tasks.py

```diff
@@ -76,174 +74,332 @@ def run(self,
-                             "created stack [%s]." % self.stack_name)
-                self.heat_client.stacks.delete(stack_id=self.stack_name)
-            elif e.suspend:
-                logger.error("Suspending unsuccessfully "
-                             "resumed stack [%s]." % self.stack_name)
-                self.heat_client.actions.suspend(stack_id=self.stack_name)
+            self.cleanup_stack(e)
 
         return stack_data
 
+    def stack_update(self, data):
+        return update_stack(self.stack_name,
+                            self.course_id,
+                            self.student_id,
+                            data)
+
+    def stack_get(self, prop=None):
+        return get_stack(self.stack_name,
+                         self.course_id,
+                         self.student_id,
+                         prop)
+
     def sleep(self):
         time.sleep(self.sleep_seconds)
 
-    def get_heat_client(self):
-        return HeatWrapper(**self.configuration).get_client()
+    def get_provider(self, name):
+        try:
+            provider = next(p for p in self.providers if p["name"] == name)
+        except StopIteration:
+            provider = None
+
+        return provider
+
+    def get_credentials(self, provider):
+        credentials = None
+        provider = self.get_provider(provider)
+        if provider:
+            credentials = get_credentials(self.settings,
+                                          provider.get("name"))
+
+        return credentials
 
-    def get_nova_client(self):
-        return NovaWrapper(**self.configuration).get_client()
+    def get_environment(self, provider):
+        provider = self.get_provider(provider)
+        return provider.get("environment")
+
+    def get_heat_client(self, provider):
+        heat_c = None
+        credentials = self.get_credentials(provider)
+        if credentials:
+            heat_c = HeatWrapper(**credentials).get_client()
+
+        return heat_c
+
+    def get_nova_client(self, provider):
+        nova_c = None
+        credentials = self.get_credentials(provider)
+        if credentials:
+            nova_c = NovaWrapper(**credentials).get_client()
+
+        return nova_c
 
     def launch_stack(self):
         """
         Launch the user stack, either by creating or resuming it.  If a reset
         is requested, delete the stack and recreate it.
 
+        """
+        logger.info("Launching stack [%s]." % self.stack_name)
+
+        # Fetch stack information from the database
+        stack = self.stack_get()
+
+        if stack.provider:
+            if self.reset:
+                self.try_provider(stack.provider, True)
+
+                stack_data = self.try_all_providers()
+            else:
+                stack_data = self.try_provider(stack.provider)
+        else:
+            stack_data = self.try_all_providers()
+
+        return stack_data
+
+    def get_provider_stack_count(self, provider):
```

Comment:
I very much like the idea that instead of keeping track of a "stacks per provider" semaphore count, we just do a `SELECT COUNT` every time we need to get the stack count. We'll need to see how this scales, I suppose, but my hunch is that this scales far better than any contentious updates.

Reply:
Yup, my feelings exactly.  However, I don't _think_ this will bog down our mariadb cluster any time soon.  The `hastexo_stack` table has not yet gone beyond 10k records in any of our production stacks, and unless we're talking tens of millions of stacks, we can probably handle our worst cases with some further InnoDB tuning.